<div style="text-align: center; background-color: #750E21; font-family: 'Trebuchet MS', Arial, sans-serif; color: white; padding: 20px; font-size: 40px; font-weight: bold; border-radius: 0 0 0 0; box-shadow: 0px 6px 8px rgba(0, 0, 0, 0.2);">
  FINAL PROJECT: RESEARCHING ON HOTEL BOOKING DEMAND 📌
</div>

<div style="text-align: center; background-color: #0766AD; font-family: 'Trebuchet MS', Arial, sans-serif; color: white; padding: 20px; font-size: 40px; font-weight: bold; border-radius: 0 0 0 0; box-shadow: 0px 6px 8px rgba(0, 0, 0, 0.2);">
  Stage 01 - Data collecting 📌
</div>

## **TABLE OF CONTENTS** 🔠 

## **INTRODUCTION TO OUR DATASET** 🚀

### **1. Overal information** 📜 

1.1 What subject is the data about ?

* The subject that our team works on is about datasets of hotel demand. The data provides insight into the business in the field of hotel booking, including 2 types: `resort hotel` and `city hotel`. 
* Using this dataset, we intend to use it for digging into hotel operations field. This research would make a contribution to capture the diverse factors impacting on hotel demand, supporting for a reasonable analysis of trends, patterns and potential development path in the hospitality industry. Eventually, we could understand more and optimize the booking processes to improve the outcomes.
* Brief on the dataset:
    * Subject area: Hospitality Management
    * More specific subject area: Revenue Management
    * Type of data: Text files and R objects
    * How data was acquired: Extraction from hotels’ Property Management System (PMS) SQL databases
    * Data format: Mixed (raw and preprocessed)
    * Experimental factors: Some of the variables were engineered from other variables from different database tables. The data point time for each observation was defined as the day prior to each booking's arrival
    * Experimental features Data was extracted via TSQL queries executed directly in the hotels’ PMS databases and R was employed to perform data analysis
    * Data source location: In **Portugal**: H1 at the resort region of Algarve and H2 at the city of Lisbon.


1.2 Source of the data:

1.3 Data license:

### **2. How the dataset was collected ?** 📜 

## **IMPORT LIBRARY** 🎄

In [2]:
import pandas
import numpy

## **RAW DATA GATHERED** 🌟 

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #FF90BC; padding: 20px; font-size: 25px; font-weight: bold; border-radius: 0 0 0 0">
  STEP 1: OBSERVE THE FIRST TABLE OF DATA H1 & H2🔥
</div>

In [3]:
data_h1 = pandas.read_csv("data/h1.csv")
data_h1.head()

IsCanceled  LeadTime  ArrivalDateYear ArrivalDateMonth  \
0           0       342             2015             July   
1           0       737             2015             July   
2           0         7             2015             July   
3           0        13             2015             July   
4           0        14             2015             July   

   ArrivalDateWeekNumber  ArrivalDateDayOfMonth  StaysInWeekendNights  \
0                     27                      1                     0   
1                     27                      1                     0   
2                     27                      1                     0   
3                     27                      1                     0   
4                     27                      1                     0   

   StaysInWeekNights  Adults  Children  ...      DepositType        Agent  \
0                  0       2         0  ...  No Deposit              NULL   
1                  0       2         0  ...  No Deposit              NULL   
2                  1       1         0  ...  No Deposit              NULL   
3                  1       1         0  ...  No Deposit               304   
4                  2       2         0  ...  No Deposit               240   

       Company DaysInWaitingList CustomerType   ADR  RequiredCarParkingSpaces  \
0         NULL                 0    Transient   0.0                         0   
1         NULL                 0    Transient   0.0                         0   
2         NULL                 0    Transient  75.0                         0   
3         NULL                 0    Transient  75.0                         0   
4         NULL                 0    Transient  98.0                         0   

   TotalOfSpecialRequests ReservationStatus ReservationStatusDate  
0                       0         Check-Out            2015-07-01  
1                       0         Check-Out            2015-07-01  
2                       0         Check-Out            2015-07-02  
3                       0         Check-Out            2015-07-02  
4                       1         Check-Out            2015-07-03  

[5 rows x 31 columns]

In [4]:
data_h1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40060 entries, 0 to 40059
Data columns (total 31 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   IsCanceled                   40060 non-null  int64  
 1   LeadTime                     40060 non-null  int64  
 2   ArrivalDateYear              40060 non-null  int64  
 3   ArrivalDateMonth             40060 non-null  object 
 4   ArrivalDateWeekNumber        40060 non-null  int64  
 5   ArrivalDateDayOfMonth        40060 non-null  int64  
 6   StaysInWeekendNights         40060 non-null  int64  
 7   StaysInWeekNights            40060 non-null  int64  
 8   Adults                       40060 non-null  int64  
 9   Children                     40060 non-null  int64  
 10  Babies                       40060 non-null  int64  
 11  Meal                         40060 non-null  object 
 12  Country                      39596 non-null  object 
 13  MarketSegment   

`COMMENT:`

In [6]:
# add a column to the data frame
data_h1["hotel_type"] = "resort hotel"
data_h1.head()

IsCanceled  LeadTime  ArrivalDateYear ArrivalDateMonth  \
0           0       342             2015             July   
1           0       737             2015             July   
2           0         7             2015             July   
3           0        13             2015             July   
4           0        14             2015             July   

   ArrivalDateWeekNumber  ArrivalDateDayOfMonth  StaysInWeekendNights  \
0                     27                      1                     0   
1                     27                      1                     0   
2                     27                      1                     0   
3                     27                      1                     0   
4                     27                      1                     0   

   StaysInWeekNights  Adults  Children  ...        Agent      Company  \
0                  0       2         0  ...         NULL         NULL   
1                  0       2         0  ...         NULL         NULL   
2                  1       1         0  ...         NULL         NULL   
3                  1       1         0  ...          304         NULL   
4                  2       2         0  ...          240         NULL   

  DaysInWaitingList CustomerType   ADR  RequiredCarParkingSpaces  \
0                 0    Transient   0.0                         0   
1                 0    Transient   0.0                         0   
2                 0    Transient  75.0                         0   
3                 0    Transient  75.0                         0   
4                 0    Transient  98.0                         0   

   TotalOfSpecialRequests  ReservationStatus ReservationStatusDate  \
0                       0          Check-Out            2015-07-01   
1                       0          Check-Out            2015-07-01   
2                       0          Check-Out            2015-07-02   
3                       0          Check-Out            2015-07-02   
4                       1          Check-Out            2015-07-03   

     hotel_type  
0  resort hotel  
1  resort hotel  
2  resort hotel  
3  resort hotel  
4  resort hotel  

[5 rows x 32 columns]

In [7]:
data_h2 = pandas.read_csv("data/h2.csv")
data_h2.head()

IsCanceled  LeadTime  ArrivalDateYear ArrivalDateMonth  \
0           0         6             2015             July   
1           1        88             2015             July   
2           1        65             2015             July   
3           1        92             2015             July   
4           1       100             2015             July   

   ArrivalDateWeekNumber  ArrivalDateDayOfMonth  StaysInWeekendNights  \
0                     27                      1                     0   
1                     27                      1                     0   
2                     27                      1                     0   
3                     27                      1                     2   
4                     27                      2                     0   

   StaysInWeekNights  Adults  Children  ...      DepositType        Agent  \
0                  2       1       0.0  ...  No Deposit                 6   
1                  4       2       0.0  ...  No Deposit                 9   
2                  4       1       0.0  ...  No Deposit                 9   
3                  4       2       0.0  ...  No Deposit                 9   
4                  2       2       0.0  ...  No Deposit                 9   

       Company DaysInWaitingList CustomerType   ADR  RequiredCarParkingSpaces  \
0         NULL                 0    Transient   0.0                         0   
1         NULL                 0    Transient  76.5                         0   
2         NULL                 0    Transient  68.0                         0   
3         NULL                 0    Transient  76.5                         0   
4         NULL                 0    Transient  76.5                         0   

   TotalOfSpecialRequests ReservationStatus ReservationStatusDate  
0                       0         Check-Out            2015-07-03  
1                       1          Canceled            2015-07-01  
2                       1          Canceled            2015-04-30  
3                       2          Canceled            2015-06-23  
4                       1          Canceled            2015-04-02  

[5 rows x 31 columns]

In [8]:
data_h2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79330 entries, 0 to 79329
Data columns (total 31 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   IsCanceled                   79330 non-null  int64  
 1   LeadTime                     79330 non-null  int64  
 2   ArrivalDateYear              79330 non-null  int64  
 3   ArrivalDateMonth             79330 non-null  object 
 4   ArrivalDateWeekNumber        79330 non-null  int64  
 5   ArrivalDateDayOfMonth        79330 non-null  int64  
 6   StaysInWeekendNights         79330 non-null  int64  
 7   StaysInWeekNights            79330 non-null  int64  
 8   Adults                       79330 non-null  int64  
 9   Children                     79326 non-null  float64
 10  Babies                       79330 non-null  int64  
 11  Meal                         79330 non-null  object 
 12  Country                      79306 non-null  object 
 13  MarketSegment   

`COMMENT:`

In [9]:
# add a column to the data frame
data_h2["hotel_type"] = "city hotel"
data_h2.head()

IsCanceled  LeadTime  ArrivalDateYear ArrivalDateMonth  \
0           0         6             2015             July   
1           1        88             2015             July   
2           1        65             2015             July   
3           1        92             2015             July   
4           1       100             2015             July   

   ArrivalDateWeekNumber  ArrivalDateDayOfMonth  StaysInWeekendNights  \
0                     27                      1                     0   
1                     27                      1                     0   
2                     27                      1                     0   
3                     27                      1                     2   
4                     27                      2                     0   

   StaysInWeekNights  Adults  Children  ...        Agent      Company  \
0                  2       1       0.0  ...            6         NULL   
1                  4       2       0.0  ...            9         NULL   
2                  4       1       0.0  ...            9         NULL   
3                  4       2       0.0  ...            9         NULL   
4                  2       2       0.0  ...            9         NULL   

  DaysInWaitingList CustomerType   ADR  RequiredCarParkingSpaces  \
0                 0    Transient   0.0                         0   
1                 0    Transient  76.5                         0   
2                 0    Transient  68.0                         0   
3                 0    Transient  76.5                         0   
4                 0    Transient  76.5                         0   

   TotalOfSpecialRequests  ReservationStatus ReservationStatusDate  hotel_type  
0                       0          Check-Out            2015-07-03  city hotel  
1                       1           Canceled            2015-07-01  city hotel  
2                       1           Canceled            2015-04-30  city hotel  
3                       2           Canceled            2015-06-23  city hotel  
4                       1           Canceled            2015-04-02  city hotel  

[5 rows x 32 columns]

🧮 Now, we will check if the two columns have the same feature columns ?

In [10]:
# compare the two datasets if they have the same columns return True/False
columns_equal = (data_h1.columns == data_h2.columns).all()
columns_equal

True

🎊 Wonderful! Both of two data samples have the same number and content of features.\
🎊 Two of data samples are just different in the number of total entries: **H1** with `40060` entries and **H2** with `79330` entries.

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #FF90BC; padding: 20px; font-size: 25px; font-weight: bold; border-radius: 0 0 0 0">
  STEP 2: DISCOVER MEANING OF THE FEATURES🔥
</div>

| Variable                          | Type       | Description          |
|:---------------------------------|:-----------:|:---------------------|
|`ADR`| Numeric| Average Daily Rate|
|`Adults`| Integer | Number of adults|
|`Agent` | Categorical | ID of the travel agency that made the booking |
|`ArrivalDateDayOfMonth`| Integer | Day of the month of the arrival date|
|`ArrivalDateMonth`| Categorical | Month of arrival date with 12 categories: "January" to "December"|
|`ArrivalDateWeekNumber`| Integer | Week number of the arrival date|
|`ArrivalDateYear`| Integer | Year of arrival date |
|`AssignedRoomType`| Categorical | Code for the type of room assigned to the booking. Sometimes the assigned room type differs from the reserved room type due to hotel operation reasons (e.g.overbooking) or by customer request.|
|`Babies`| Integer | Number of babies|
|`BookingChanges`| Integer | Number of changes made to the booking from the moment the booking was entered on the PMS until the moment of check-in or cancellation|
|`Children`| Integer | Number of children|
|`Company` | Categorical | ID of the company that made the booking or responsible for paying the booking (ID is presented for anonymity reasons)|
|`Country`| Categorical | Country of origin |
|`CustomerType`| Categorical | Type of booking, in 4 types: *Contract*, *Group*, *Transient* and *Transient-party*|
|`DaysInWaitingList`| Integer | Number of days the booking was in the waiting list before it was confirmed to the customer.|
|`DepositType`| Categorical | Indication on if the customer made a deposit to guarantee the booking, including 3 categories: *No Deposit*, *Non Refund*, *Refundable*|
|`DistributionChannel`| Categorical | Booking distribution channel: **TA** (*Travel Agents*) and **TO** (*Tour Operators*)|
|`IsCanceled`| Categorical | Value indicating if the booking was canceled **(1)** or not **(0)**|
|`IsRepeatedGuest`| Categorical | Value indicating if the booking name was from a repeated guest **(1)** or not **(0)**|
|`LeadTime`| Integer | Number of days that elapsed between the entering date of the booking into the PMS and the arrival date|
|`MarketSegment`|Integer| Market segment designation: **TA** (*Travel Agents*) and **TO** (*Tour Operators*)|
|`Meal`| Categorical | Type of meal booked: **Undefined/SC** (*no meal package*), **BB** (*Bed & Breakfast*), **HB** (*breakfast and one other meal*) and **FB** (*breakfast, lunch and dinner*)|
|`PreviousBookingsNotCanceled`|Integer| Number of previous bookings not cancelled by the customer prior to the current booking|
|`PreviousCancellations`|Integer| Number of previous bookings that were cancelled by the customer prior to the current booking|
|`RequiredCardParkingSpaces`|Integer|Number of car parking spaces required by the customer|
|`ReservationStatus`|Categorical|Reservation last status, assuming one of 3 categories: **Canceled**, **Check-Out** and **No-Show**|
|`ReservationStatusDate`|Date| Date at which the last status was set.|
|`ReservedRoomType`| Categorical | Code of room type reserved.|
|`StaysInWeekendNights`| Integer | Number of weekend nights (Saturday or Sunday) the guest stayed or booked to stay at the hotel|
|`StaysInWeekNights`| Integer | Number of week nights (Monday to Friday)|
|`TotalOfSpecialRequests`| Integer | Number of special requests made by the customer|


<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #FF90BC; padding: 20px; font-size: 25px; font-weight: bold; border-radius: 0 0 0 0">
  STEP 3: MERGE THE TWO DATA SAMPLE🔥
</div>

In [12]:
raw_data = pandas.concat([data_h1, data_h2], axis=0).reset_index(drop=True)
raw_data.sample(10)

IsCanceled  LeadTime  ArrivalDateYear ArrivalDateMonth  \
60628            1       496             2016         November   
91218            0        18             2016             June   
100903           0        36             2016         November   
47704            1       116             2016         February   
98416            0         3             2017          January   
102332           0        56             2016         November   
35383            0        14             2017            April   
105581           0        24             2017         February   
14566            0         1             2015        September   
5681             1       203             2016              May   

        ArrivalDateWeekNumber  ArrivalDateDayOfMonth  StaysInWeekendNights  \
60628                      47                     17                     0   
91218                      25                     13                     1   
100903                     45                      1                     0   
47704                      10                     28                     2   
98416                       2                     12                     0   
102332                     48                     24                     0   
35383                      16                     19                     0   
105581                      6                      5                     2   
14566                      36                      5                     1   
5681                       20                     10                     0   

        StaysInWeekNights  Adults  Children  ...        Agent      Company  \
60628                   2       2       0.0  ...            1         NULL   
91218                   4       1       0.0  ...            7         NULL   
100903                  3       2       0.0  ...           28         NULL   
47704                   2       2       0.0  ...          134         NULL   
98416                   2       1       0.0  ...         NULL         NULL   
102332                  3       2       0.0  ...           14         NULL   
35383                   2       1       0.0  ...          104         NULL   
105581                  5       1       0.0  ...         NULL          407   
14566                   1       2       2.0  ...          240         NULL   
5681                    2       2       0.0  ...          248         NULL   

       DaysInWaitingList     CustomerType     ADR  RequiredCarParkingSpaces  \
60628                  0        Transient   62.80                         0   
91218                  0        Transient   92.40                         0   
100903                 0        Transient   75.00                         0   
47704                  0  Transient-Party   72.00                         0   
98416                  0        Transient   75.00                         1   
102332                 0  Transient-Party   95.38                         0   
35383                  0        Transient   65.00                         0   
105581                 0  Transient-Party   75.00                         0   
14566                  0        Transient  178.50                         1   
5681                   0  Transient-Party   46.00                         0   

        TotalOfSpecialRequests  ReservationStatus ReservationStatusDate  \
60628                        0           Canceled            2015-10-21   
91218                        1          Check-Out            2016-06-18   
100903                       0          Check-Out            2016-11-04   
47704                        0            No-Show            2016-02-28   
98416                        1          Check-Out            2017-01-14   
102332                       1          Check-Out            2016-11-27   
35383                        0          Check-Out            2017-04-21   
105581                       0          Check-Out            2017-02-12   
14566             

In [13]:
raw_data.to_csv("./data/raw_data/raw_data.csv", index=False)